Add Further Selection Functionality for Regions2D

ctuguinay opened the pull request:
### Add Further Selection Functionality for Regions2D

This PR addresses #93.

- Added selection functionality for depth range.
- Added selection functionality for time range.
- Created a test to check for both time and depth functionality.

leewujung commented:
Comments from reviewing #95 since it is related to the region selection:
for the below section:
https://github.com/OSOceanAcoustics/echoregions/blob/f9f4c36418700e6e7e8a9c144542de630cc84eb7/echoregions/regions2d/regions2d.py#L131-L154
- When the input `region` is a dataframe or a series, the existence of the field "region_id" should be checked. 
- When the input `region` is None, a warning should be issued so that people know they have not sub-selected anything

However, I actually think that we should not allow this much variability: What about we only allow `int`, `float`, `string`, or a list of these? That's much cleaner and still easy to do for users if they have a dataframe or a series containing the desired region_id.

ctuguinay commented:
> Comments from reviewing #95 since it is related to the region selection: for the below section:
> 
> https://github.com/OSOceanAcoustics/echoregions/blob/f9f4c36418700e6e7e8a9c144542de630cc84eb7/echoregions/regions2d/regions2d.py#L131-L154
> 
> * When the input `region` is a dataframe or a series, the existence of the field "region_id" should be checked.
> * When the input `region` is None, a warning should be issued so that people know they have not sub-selected anything
> 
> However, I actually think that we should not allow this much variability: What about we only allow `int`, `float`, `string`, or a list of these? That's much cleaner and still easy to do for users if they have a dataframe or a series containing the desired region_id. Doing this will require all usages of `select_region` to be checked to make sure they still function as intended (and a good excuse to write more tests!).

@leewujung Just sent a push addressing this!

leewujung commented on echoregions/regions2d/regions2d.py:
```diff
@@ -73,42 +73,140 @@ def to_json(self, save_path: str = None) -> None:
         Parameters
         ----------
-        region : float, str, list, Series, DataFrame, ``None``
+        region_id : float, int, str, list, ``None``
             A region id provided as a number, string, list of these,
             or a DataFrame/Series containing the region_id column name.
```
```suggestion
        region_id : float, int, str, list, ``None``
            A region id provided as a number, a string, or list of these.
```

leewujung commented on echoregions/regions2d/regions2d.py:
```diff
@@ -73,42 +73,140 @@ def to_json(self, save_path: str = None) -> None:
+        copy=False,
     ) -> DataFrame:
-        """Ensure that region is a DataFrame.
+        """Selects a subset of this region object's dataframe.
```
```suggestion
        """Selects a subset of this Region2D object's dataframe.
```

leewujung commented on echoregions/regions2d/regions2d.py:
```diff
@@ -73,42 +73,140 @@ def to_json(self, save_path: str = None) -> None:
         DataFrame
             A DataFrame subselected from Regions2D.data.
-            There is a row for each region id provided by the region parameter.
+            There is a row for each region id provided by the region_id parameter,
+            and each row has time and depth within or on the boundaries passed
+            in by the time_range and depth_range values.
```
Just some formatting stuff
```suggestion
            There is a row for each region id provided by the ``region_id`` parameter,
            and each row has time and depth within or on the boundaries passed
            in by the ``time_range`` and ``depth_range`` values.
```

leewujung commented on echoregions/regions2d/regions2d.py:
```diff
@@ -73,42 +73,140 @@ def to_json(self, save_path: str = None) -> None:
+        untouched = True
+        if region_id is not None:
+            if (
+                isinstance(region_id, float)
+                or isinstance(region_id, int)
+                or isinstance(region_id, str)
```
Can just be `isinstance(region_id, (float, int, str))`?

leewujung commented on echoregions/regions2d/regions2d.py:
```diff
@@ -73,42 +73,140 @@ def to_json(self, save_path: str = None) -> None:
+                        time_range[1], Timestamp
+                    ):
+                        if time_range[0] < time_range[1]:
+                            if region is None:
+                                region = self.data
```
This assignment is hidden and individually assigned in each of the blocks when you test for `region_id`, `time_range` and `depth_range`. I recommend that you move this out and do the assignment upfront (probably with `.copy()` so that you do not modify `self.data` by default) and do modification in each block so that it is easier to keep track of what's what. This way you do not need `untouched` either -- if it is untouched, just assign it back to `self.data`. The dataframe is small in general so I wouldn't be too worried about efficiency just yet.

ctuguinay replied:
I still think they should have the option to modify the actual inner dataframe, so I'll still have that if copy is false, then we have that the selection modifies the inner dataframe, and as such the class. But I'll set the copy as True so the user does not automatically do this, and instead just receives a copy of the inner dataframe.

leewujung replied:
Yeah I wasn't suggesting removing that option, so your changes look good to me!

leewujung commented on echoregions/regions2d/regions2d.py:
```diff
@@ -73,42 +73,140 @@ def to_json(self, save_path: str = None) -> None:
+                        depth_range[1], (float, int)
+                    ):
+                        if depth_range[0] < depth_range[1]:
+                            if region is None:
+                                region = self.data
+                            for index, row in region.iterrows():
+                                remove_row = False
+                                for depth in row["depth"]:
+                                    if depth_range[0] > depth or depth_range[1] < depth:
+                                        remove_row = True
+                                if remove_row:
+                                    region.drop(index, inplace=True)
```
Similar comments as before for `time_range` re .apply` function and use `regions` assigned upfront.

leewujung reviewed (changes requested):
Thanks for the PR @ctuguinay! A few comments to improve readability and tighten things up. Otherwise this looks good!

Also for a small PR like this you could respond individually to each comment, or "resolve" them once you're done with the changes with that button.

ctuguinay commented:
Hey @leewujung, I've addressed your latest review comments:

- I've reformatted and cleaned up the comments, descriptions, and `isinstance` code.
- I've instantiated regions upfront and removed extra fluff associated with not instantiating regions upfront
- I've replaced the selection of rows using an itterrows implementation with a df.apply and lambda implementation

leewujung commented:
LGTM! Feel free to merge this!
